api/client/build.go

```diff
@@ -390,36 +278,209 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+		// use it if we found it, otherwise ignore this check
+		if _, err = os.Lstat(dockerfileAbspath); os.IsNotExist(err) {
+			altPath := filepath.Join(absBase, strings.ToLower(api.DefaultDockerfileName))
+			if _, err = os.Lstat(altPath); err == nil {
+				dockerfileAbspath = altPath
+			}
+		}
+	}
+
+	// Verify that 'filename' is within the build context
+	dockerfileAbspath, err = symlink.FollowSymlinkInScope(dockerfileAbspath, absBase)
+	if err != nil {
+		return fmt.Errorf("The Dockerfile (%s) must be within the build context (%s)", origDockerfile, baseDir)
+	}
+
+	if _, err := os.Lstat(dockerfileAbspath); err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("Cannot locate Dockerfile: baseDir: %q, origDockerfile: %q, absDockerfile: %q", baseDir, origDockerfile, dockerfileAbspath)
+		}
+		return fmt.Errorf("unable to stat Dockerfile: %v", err)
+	}
+
+	if *dockerfileName, err = filepath.Rel(absBase, dockerfileAbspath); err != nil {
+		return fmt.Errorf("unable to get relative Dockerfile path: %v", err)
+	}
+
+	return nil
+}
+
+func createDockerfileFromReader(r io.Reader, filename string) error {
+	dockerfile, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.FileMode(0600))
+	if err != nil {
+		return fmt.Errorf("unable to create temporary Dockerfile: %v", err)
+	}
+	defer dockerfile.Close()
+
+	if _, err := io.Copy(dockerfile, r); err != nil {
+		return fmt.Errorf("unable to write temporary Dockerfile: %v", err)
+	}
+
+	return nil
+}
+
+// As a special case, 'docker build -' will build from either an empty context
+// with the contents of stdin as a Dockerfile, or a tar archive of context from
+// stdin. getContextFromStdin will create a local temporary directory which
+// contains this context and/or Dockerfile.
+func (cli *DockerCli) getContextFromStdin(dockerfileName *string) (tempDir string, err error) {
+	buf := bufio.NewReader(cli.in)
+
+	magic, err := buf.Peek(tarHeaderSize)
+	if err != nil && err != io.EOF {
+		return "", fmt.Errorf("failed to peek context header from STDIN: %v", err)
+	}
+
+	if tempDir, err = ioutil.TempDir("", "docker-build-context-"); err != nil {
+		return "", fmt.Errorf("unbale to create temporary context directory: %v", err)
+	}
+
+	defer func(d string) {
+		if err != nil {
+			os.RemoveAll(d)
+		}
+	}(tempDir)
+
+	if !archive.IsArchive(magic) { // Input should be read as a Dockerfile.
+		// -f option has no meaning when we're reading it from stdin,
+		// so just use our default Dockerfile name
+		*dockerfileName = api.DefaultDockerfileName
+
+		return tempDir, createDockerfileFromReader(buf, filepath.Join(tempDir, *dockerfileName))
+	}
+
+	if err := archive.Untar(buf, tempDir, nil); err != nil {
+		return "", fmt.Errorf("unable to extract stdin to temporary context direcotry: %v", err)
+	}
+
+	return tempDir, getDockerfileRelPath(dockerfileName, tempDir)
+}
+
+func (cli *DockerCli) getContextFromGitURL(gitURL string, dockerfileName *string) (tempDir string, err error) {
```

Comment:
I'm a little worried that this struct method set is sprawling. There is some sort of context here that is being created as a directory. This seems orthogonal to the concerns of the `DockerCli` type. These operations are related to building.